python/arcticdb/util/utils.py

```diff
@@ -0,0 +1,325 @@
+"""
+Copyright 2024 Man Group Operations Limited
+Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+from arcticdb import Arctic
+from tests.stress.arcticdb.version_store.test_stress_finalize_stage_data import CachedDFGenerator, stage_chunks
+from arcticdb.util.utils import TimestampNumber
+"""
+
+from typing import List
+from typing_extensions import Literal, Tuple, Union
+import numpy as np
+import pandas as pd
+
+from arcticdb.util.test import create_datetime_index, get_sample_dataframe
+from arcticdb.version_store.library import Library
+
+class  TimestampNumber:
+    """
+        Represents the timestamp as a typed number (can be seconds, minutes, hours),
+        Thus allowing simple arithmetic with numbers - adding and subtracting
+        specified number of same type units results in increasing or decreasing the timestamp with same
+        amount of time that type/freq represents
+
+        In other words any numbers added, subtracted or compared with this type
+        are implicitly considered as instances of the same type seconds, minutes, hours
+        and operations are carried on naturally.
+
+        Supported are int operation for increment and decrement
+
+        For comparison you can do it with any number or Timestamp object
+
+        0 is Timestamp(0) etc
+    """
+    SupportedFreqTypes = Literal['s','m','h']
+    
+    DEFAULT_FREQ:SupportedFreqTypes = 's'
+
+    TIME_ZERO : pd.Timestamp = pd.Timestamp(0)
+
+    def __init__(self, value:np.int64, type:SupportedFreqTypes=DEFAULT_FREQ) -> None: 
+        self.init_value:np.int64 = value
+        self.value:np.int64 = value
+        self.__type:TimestampNumber.SupportedFreqTypes = type
+
+
+    def get_type(self) -> SupportedFreqTypes:
+        return self.__type
```

Comment:
__ is also for private fields - ie fields that you would not expose externally (like read only fields). As far as I know python makes them harder to access by adding something like "self.__private__type" to simulte private fields. Thus in this particular case if we allow to access and modify __type this will jeopordize the object as if we assign directly here "h" and it was "s" the TimeStmapNumber now is totally different. It would require full conversion of the whole object, which is possible but not needed at this point of usage of this class

Reply:
The convention is to have a single underscore for private fields, not two